File: 1306-jump-game-iii/1306-jump-game-iii.php
class Solution {

    /**
     * @param Integer[] $arr
     * @param Integer $start
     * @return Boolean
     */

    function canReach($arr, $start) {
        $n = count($arr);
        $array = array();
        $dp = function($i) use (&$dp, $arr, $n, &$array){
            if ($i < 0 || $i >= $n){
                return False;
            }
            if($arr[$i] == 0){
                return True;
            }
            if (isset($array[$i])){
                return False;
            }
            $array[$i] = True;
            $left = $dp($i + $arr[$i]);
            $right = $dp($i - $arr[$i]);
            if ($left || $right){
                return True;
            }
            return False;
        };
        return $dp($start);
        
    }
}

File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        n = len(arr)
        seen = set()

        @cache
        def dp(i):
            
            if i < 0 or i >= n:
                return False
            if arr[i] == 0:
                return True
            if i in seen:
                return False
            seen.add(i)
            left = dp(i - arr[i])
            right = dp(i + arr[i])
            if left or right:
                return True

            return False
        return dp(start)